Fixed #54: added search result processing to add facet aggregates

grtjn opened the pull request:
Fixes #54

joemfb commented on src/ml-search.service.js:
```diff
@@ -1225,6 +1233,77 @@
+            // these work for all index types
+            myOption.aggregate = [
+              { apply: 'count' },
+              { apply: 'min' },
```
I'm pretty sure `min`/`max` don't work for all index types. I tested pretty extensively when I was developing https://github.com/joemfb/analytics-query-browser. Here's the module I used to create the config for that project: https://gist.github.com/joemfb/b682504c7c19cd6fae11

grtjn replied:
I admit I didn't test *all* types, but I did test xs:string, xs:dateTime, and geo-elem-pair. I think that covers the most difficult ones..

grtjn replied:
Looks like you are only missing string in your list for min/max. And I'd expect anyUri to work as well..

joemfb commented:
This makes sense. Let me review my values branch, check the tests, etc. Than I'll do a PR. Once that's merged, you can rebase on top of this. Should let you simplify parts of it.

grtjn commented:
Looking forward to your PR.. :)

grtjn commented:
I rebased, synched my getAggregates code with your new showMoreFacets. I ran tests against a project I was using for the new functionality, and found a few issues in values and showMoreFacets. The options need to be put in combined.search.options, and you need to cloneDeep constraints from storedOptions. Otherwise you end up messing them up..

Looks good like this? gulp tests run clean. I can squash the commits if you like..

joemfb commented on src/ml-search.service.js:
```diff
@@ -767,13 +769,15 @@
-        newOptions.values['values-option'] = newOptions.values.range['facet-option'];
+        constraint = _.cloneDeep(constraint);
+
+        var newOptions = { constraint: _.cloneDeep(constraint), values: constraint };
```
Are both `_.cloneDeep()` invocations necessary? The original constraint object isn't used again, so I would think one would be enough ...

joemfb commented:
My only question is the `cloneDeep` calls. Otherwise, I'd say squash the commits and we can merge (I don't think they're sufficiently differentiated to be valuable separately).

Looks good!

grtjn commented:
two clones was indeed overkill. I fixed that, and squashed the commits!

joemfb commented:
Actually, I was wrong; there's one more thing...

Could you write some tests for the `getAggregates` method? There's several examples of tests for similar methods. You can mock the data, and then use the code coverage report to figure out where the gabs are, and what additional tests you need.

grtjn commented:
I wish it was that easy as copy paste one of the other tests, but the code loops over all facets, and makes all kinds of backend calls. I'm not very good at writing tests for that. Can you give a few more hints how to best approach this?

joemfb commented:
I've tried to make the test harness straightforward, but some of this code is just really hard to test. You can mock the dependencies directly, and just make sure that they're invoked with the correct parameters, or you can use `$httpBackend` to mock the underlying HTTP requests.

`showMoreFacets` is one of the more complicated sets of tests, and probably a useful example here: https://github.com/joemfb/ml-search-ng/blob/master/test/spec/ml-search.service.js#L580

Also, I mocked MLRest directly here: https://github.com/joemfb/ml-search-ng/blob/master/test/spec/ml-search.service.js#L1266. That might be a simpler approach in this case.

grtjn commented:
Thanks for the hints, I added two test cases for getAggregates. I also rebased, and fixed the test cases for showMoreFacets (issue #45). Additionally, jshint was giving some suggestions at my end, to good use I think..

I'll squash commits if you like the changes..

joemfb commented:
This looks good, thanks for adding the test cases!

Please do rebase down to one commit, and, when you do, can you remove `dist/*`? I want those to change only with tagged release commits.

grtjn commented:
Whoops, didn't mean to commit dist. I have squashed the commits..

joemfb commented on src/ml-search.service.js:
```diff
@@ -1137,16 +1141,16 @@
-      params.start = params.start || 1;
-      params.limit = params.limit || 20;
+      params.start = params.start !== undefined ? params.start : 1;
+      params.limit = params.limit !== undefined ? params.limit : 20;
```
This more explicit check is to allow falsy values as parameters (such as `limit=0`)?

grtjn replied:
Yes exactly. The getAggregates passes through limit=0 to only get aggregates, and not bother about values. My test case was complaining about it.. ;-)
